### ISSUE TITLE
2.4 pre-installer does not source nix after reboot in zsh

### ISSUE THREAD

**Issue by SimonTheLeg** [NixOS/nix#5406]
**Describe the bug**

After an installation using the 2.4pre installer on my new laptop, everything works fine at first, but after a reboot nix is not sourced correctly. 

Using `zsh -o SOURCE_TRACE`, I think I found the culprit: 

_On my old laptop_
```sh
❯ zsh -o SOURCE_TRACE
+/etc/zshenv:1> <sourcetrace>
+/nix/var/nix/profiles/default/etc/profile.d/nix-daemon.sh:1> <sourcetrace>
```

_But on the new one /etc/zshenv does not exist_

```sh
simonbein@Simons-MBP-2 ~ % ls /etc/zshenv
ls: /etc/zshenv: No such file or directory
```

We can see this in the `zsh -o SOURCE_TRACE` as well
```sh
simonbein@Simons-MBP-2 dotfiles % zsh -o SOURCE_TRACE
+/Users/simonbein/.zshenv:1> <sourcetrace>
+/etc/zshrc:1> <sourcetrace>
```

So my current assumption is that the old installer did edit `/etc/zshenv`, but a newer one does not. I have to admit though I have ran quite a lot of different nix installers on the old macbook, so I have no idea which one of the older versions did that.

**Steps To Reproduce**

1. Install using the 2.4pre-installer
```sh
curl -L https://releases.nixos.org/nix/nix-2.4pre-rc1/install | sh
```
2. Reboot the system
3. Try any command sourced by nix or even nix itself
```sh
nix
zsh: command not found: nix
```

**Workaround I use for now**
Currently I simply created the `/etc/zshenv` manually with the following content:

```sh
# Nix
if [ -e '/nix/var/nix/profiles/default/etc/profile.d/nix-daemon.sh' ]; then
  . '/nix/var/nix/profiles/default/etc/profile.d/nix-daemon.sh'
fi
# End Nix
```

**Expected behavior**

Nix is correctly sourced when starting a new zsh

**`nix-env --version` output**

Now this of course returns command not found ;)
But: 
```sh
/nix/var/nix/profiles/default/bin/nix-env --version
nix-env (Nix) 2.4pre-rc1
```

**Additional context**

- macOS 11.6
- brand new laptop

**Comment by jsoo1** [comment]
@abathur This is a problem when the machine is a substituter or remote builder, or indeed, you just want to execute a nix command remotely via ssh. I ran into this problem when on another host I ran `nix store ping --store ssh://user@example`. The problem with the `example` host is that `etc/zshrc` is not sourced for non-login shells. This breaks all kinds of commands like `nix-copy-closure` `nix store ping` and more. Even `ssh user@example nix-env -q` doesn't work because of the sourcing location. I'm afraid nix might need to be a bit more sophisticated in where it is sourcing the profile.

**Comment by jsoo1** [comment]
I am not sure if this is 100% a darwin problem, but I am reading the discussion in https://github.com/NixOS/nix/issues/4169 and getting sad, again

**Comment by abathur** [comment]
> I'm afraid nix might need to be a bit more sophisticated in where it is sourcing the profile.

I agree (though I imagine it's easier said than done). I've been [fishing around](https://discourse.nixos.org/t/installer-test-suite-small-project-s-high-leverage-help-wanted/13662/3) for anyone who can figure out the basics of a proper test suite--and flushing out shell profile problems and regressions is, in my opinion, one of the most-important reasons to have it. 

As an observer, I think there's a lot of variety in the wild, and it doesn't seem like many (any?) people have a strong high-altitude cross-platform understanding. It's common enough for someone to touch the ~profile code because something isn't working for them without realizing their setup isn't normative/universal and break other users in the process.

> The problem with the example host is that /etc/zshrc is not sourced for non-login shells. ... Even ssh user@example nix-env -q doesn't work because of the sourcing location. 

I suppose the effect is the same, but after reading `man zsh` on Catalina I think this is actually because the remote shell isn't interactive. Here's the section:

> STARTUP/SHUTDOWN FILES
> 
> Commands are first read from /etc/zshenv; this cannot be overridden. Subsequent behaviour is modified by the RCS and GLOBAL_RCS options; the former affects all startup files, while the second only affects global startup files (those shown here with an path starting with a /). If one of the options is unset at any point, any subsequent startup file(s) of the corresponding type will not be read. It is also possible for a file in $ZDOTDIR to re-enable GLOBAL_RCS. Both RCS and GLOBAL_RCS are set by default.
> 
> Commands are then read from $ZDOTDIR/.zshenv. If the shell is a login shell, commands are read from /etc/zprofile and then $ZDOTDIR/.zprofile. Then, if the shell is interactive, commands are read from /etc/zshrc and then $ZDOTDIR/.zshrc. Finally, if the shell is a login shell, /etc/zlogin and $ZDOTDIR/.zlogin are read.

**Comment by jsoo1** [comment]
> I agree (though I imagine it's easier said than done). I've been [fishing around](https://discourse.nixos.org/t/installer-test-suite-small-project-s-high-leverage-help-wanted/13662/3) for anyone who can figure out the basics of a proper installer test suite--and flushing out shell profile problems and regressions is, in my opinion, one of the most-important reasons to have it.

A test matrix for installers would be a massive help, wouldn't it?

> As an observer, I think there's a lot of variety in the wild, and it doesn't seem like many (any?) people have a strong high-altitude cross-platform understanding. It's common enough for someone to touch the ~profile code because something isn't working for them without realizing their setup isn't normative/universal and break other users in the process.

Even more reason to setup a test matrix.

> I suppose the effect is the same, but after reading `man zsh` on Catalina I think this is actually because the remote shell isn't interactive.

Yes, you're right. I tried both /etc/zprofile and /etc/zshenv which may be why I mentioned login instead of interactive shell.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
